app/internal/features.py

```diff
@@ -0,0 +1,209 @@
+from fastapi import Depends, Request
+from functools import wraps
+from starlette.responses import RedirectResponse
+from typing import List
+
+from app.database.models import UserFeature, Feature, User
+from app.dependencies import get_db, SessionLocal
+from app.internal.security.dependencies import current_user_from_db
+from app.internal.security.ouath2 import get_authorization_cookie
+from app.internal.features_index import features
+from app.internal.utils import create_model
+
+
+def feature_access_filter(call_next):
+    @wraps(call_next)
+    async def wrapper(*args, **kwargs):
+        request = kwargs["request"]
+
+        if request.headers["user-agent"] == "testclient":
+            # in case it's a unit test.
+            return await call_next(*args, **kwargs)
+
+        # getting the url route path for matching with the database.
+        route = "/" + str(request.url).replace(str(request.base_url), "")
+
+        # getting access status.
+        access = await is_access_allowd(route=route, request=request)
+
+        if access:
+            # in case the feature is enabled or access is allowed.
+            return await call_next(*args, **kwargs)
+
+        elif "referer" not in request.headers:
+            # in case request come straight from address bar in browser.
+            return RedirectResponse(url="/")
+
+        # in case the feature is disabled or access isn't allowed.
+        return RedirectResponse(url=request.headers["referer"])
+
+    return wrapper
+
+
+def create_features_at_startup(session: SessionLocal) -> bool:
+    for feat in features:
+        if not is_feature_exists(feature=feat, session=session):
+            create_feature(**feat, db=session)
+    return True
+
+
+def is_association_exists_in_db(
+    form: dict,
+    session: SessionLocal,
+    user: User,
+) -> bool:
+    db_association = (
+        session.query(UserFeature)
+        .filter_by(feature_id=form["feature_id"], user_id=user.id)
+        .first()
+    )
+
+    return db_association is not None
+
+
+def delete_feature(
+    feature: Feature,
+    session: SessionLocal = Depends(get_db),
+) -> None:
+    session.query(UserFeature).filter_by(feature_id=feature.id).delete()
+    session.query(Feature).filter_by(id=feature.id).delete()
+    session.commit()
+
+
+def is_feature_exists(feature: dict, session: SessionLocal) -> bool:
```

Comment:
https://stackoverflow.com/questions/7646173/sqlalchemy-exists-for-query

Reply:
I implemented what you sent all over the code you'll see it in the next push :)
That's much better...